src/main/main.c

```diff
@@ -631,68 +698,63 @@ int fd_listen(re_sock_t fd, int flags, fd_h *fh, void *arg)
 			return err;
 	}
 
+	err = fhs_update(re, &fhs, fd, flags, fh, arg);
+	if (err)
+		return err;
+
+	switch (re->method) {
+#ifdef HAVE_SELECT
+	case METHOD_SELECT:
 #ifdef WIN32
-	/* Windows file descriptors do not follow POSIX standard ranges. */
-	i = lookup_fd_index(re, fd);
-	if (i < 0) {
-		DEBUG_WARNING("fd_listen: fd=%d - no free fd_index\n", fd);
-		return EMFILE;
-	}
+		if (re->nfds >= DEFAULT_MAXFDS)
+			err = EMFILE;
 #else
-	i = fd;
+		if (fd + 1 >= DEFAULT_MAXFDS)
+			err = EMFILE;
+#endif
+		break;
 #endif
-
-	if (i >= re->maxfds) {
-		if (flags) {
-			DEBUG_WARNING("fd_listen: fd=%d flags=0x%02x"
-				      " - Max %d fds\n",
-				      fd, flags, re->maxfds);
-		}
-		return EMFILE;
-	}
-
-	/* Update fh set */
-	if (re->fhs) {
-		re->fhs[i].fd    = fd;
-		re->fhs[i].flags = flags;
-		re->fhs[i].fh    = fh;
-		re->fhs[i].arg   = arg;
-	}
-
-	re->nfds = max(re->nfds, i+1);
-
-	switch (re->method) {
 
 #ifdef HAVE_POLL
 	case METHOD_POLL:
-		err = set_poll_fds(re, fd, flags);
+		err = set_poll_fds(re, fhs);
 		break;
 #endif
 
 #ifdef HAVE_EPOLL
 	case METHOD_EPOLL:
 		if (re->epfd < 0)
 			return EBADFD;
-		err = set_epoll_fds(re, fd, flags);
+		err = set_epoll_fds(re, fhs);
 		break;
 #endif
 
 #ifdef HAVE_KQUEUE
 	case METHOD_KQUEUE:
-		err = set_kqueue_fds(re, fd, flags);
+		err = set_kqueue_fds(re, fhs);
 		break;
 #endif
 
 	default:
 		break;
 	}
 
-	if (err) {
-		if (flags && fh) {
-			fd_close(fd);
-			DEBUG_WARNING("fd_listen: fd=%d flags=0x%02x (%m)\n",
-				      fd, flags, err);
-		}
+#ifndef WIN32
+	if (!err)
+		re->max_fd = max(re->max_fd, fd + 1);
+#endif
+
+	if (!flags) {
+		if (!re->fhs_reuse)
+			list_append(&re->fhs_delete, &fhs->le_delete, fhs);
```

Comment:
I think, this needs to check whether the event loop is running and only use `fhs_delete` if it does. If the event loop is not running, you can delete `fhs` straight away. This way you'll not leak memory as you add/remove fds while the event loop is not running.

Also, I think you're missing a remove from `fhl`.

Reply:
Is fixed now, removing from fhl is done by mem destructor.

Reply:
If `fhs` is left in `fhl` after `fd_close` then the ABA problem remains:

1. Event loop waits on fd.
2. `fd_close(fd)` is called. `fhs` is marked as invalid but remains in `fhl`.
3. `fd_listen(fd)` is called. `fhs` is unmarked.
4. Event loop wakes up to process event on the old fd. `fhs` is not marked as removed, so the protection doesn't work.

`fd_listen` must always create a new `fhs` entry, unless it is in `fhl` and *is valid*. Better just to not keep invalid `fhs` entries in `fhl` in the first place.

Reply:
I can add a check that fd_close/fd_listen should only used single threaded (main use case) with `fhs_reuse` active.

Reply:
The problem is not limited to `fhs_reuse` active. I'm most interested in the case when it is disabled.